lib/pavilion/scriptcomposer.py

```diff
@@ -0,0 +1,697 @@
+#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, 
+#  OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY 
+#  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR 
+#  TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT 
+#  OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY 
+#  OF SUCH DAMAGE.
+#
+#  ###################################################################
+
+from __future__ import unicode_literals
+import os, datetime, grp, pwd, stat
+from collections import OrderedDict
+
+""" Class to allow for scripts to be written for other modules.
+    Typically, this will be used to write bash or batch scripts. 
+"""
+
+
+class scriptHeader( object ):
+    """Class to serve as a struct for the script header."""
+
+    @property
+    def shell_path( self ):
+        """Function to return the value of the internal shell path variable."""
+        return self._shell_path
+
+    @shell_path.setter
+    def shell_path( self, value ):
+        """Function to set the value of the internal shell path variable."""
+        if value != None and not isinstance(value, unicode):
+            error = "Shell Path must be of type 'str', not {}".format(
+                    type( value ) )
+            raise TypeError( error )
+
+        self._shell_path = value
+
+    @property
+    def scheduler_macros( self ):
+        """Function to return the value of the internal scheduler macros
+        variable.
+        """
+        return self._scheduler_macros
+
+    @scheduler_macros.setter
+    def scheduler_macros( self, value ):
+        """Function to set the value of the internal scheduler macros
+        variable.
+        """
+        if value != None and not isinstance(value, OrderedDict):
+            error = "Scheduler Macro must be of type 'OrderedDict', not " +\
+                    "{}".format( type( value ) )
+            raise TypeError( error )
+
+        self._scheduler_macros = value
+
+    def __init__( self, shell_path=None, scheduler_macros=None ):
+        """Function to set the header values for the script.
+        :param string shell_path: Shell path specification.  Typically
+                                  '/bin/bash'.  default = None.
+        :param OrderedDict scheduler_macros: Scheduler macros.  If there are
+                                             elements that are only one entry,
+                                             the value will just be None.
+                                             default = None.
+        """
+        self.shell_path = shell_path
+        self.scheduler_macros = scheduler_macros
+
+    def reset( self ):
+        """Function to reset the values of the internal variables back to
+        None.
+        """
+        self.__init__()
+
+
+class scriptModules( object ):
+    """Class to serve as a struct for the script modules."""
+
+    @property
+    def explicit_specification( self ):
```

Comment:
Ok, this is the second thing so far where the name leaves me clueless about what this is.

Reply:
Explicit_specification is where you explicitly specify the module commands.  If you don't want to use the **purge**, **swaps**, **unloads**, and **loads** functions but have a more complex setup that requires a different ordering, then you can specify the commands using this function.

Reply:
How will that be exposed? We don't have any way in the config files? I guess you could do something with it in a scheduler plugin, but why would a scheduler plugin care? They should probably work without a module system anyway.